### DISCUSSION TITLE
[Proposal] Use the Smithy IDL as the source of truth for the BSP

### DISCUSSION THREAD

**Discussion by Baccata** [build-server-protocol/build-server-protocol#499] [Proposals]
## The context

Hi all. For those who don't know me, I'll introduce myself and give some context : my name is Olivier Mélois, I'm a principal engineer at Disney Streaming. I'm also an open-source maintainer and contributor, and am quite passionate about tooling. I have contributed some non-trivial stuff to mill and ammonite, and have some minor contributions to scala-cli and Intellij under my belt.

I maintain several open-source Scala libraries under the Disneystreaming flag :

* https://github.com/disneystreaming/weaver-test (a test framework)
* https://github.com/disneystreaming/smithy4s/  (a code-generation tool )

I also share a github organisation with my ex-colleague and friend, Anton Sviridov, which we called [neandertech](https://github.com/neandertech/), and where we maintain a small ecosystem related to the LSP, written in pure Scala : see [langoustine](https://github.com/neandertech/langoustine). 

I'm saying all this to give some context and show that I'm not (just) a crazy person coming to the repo with random ideas, but I like to think I put my money where my mouth is.

## The proposal

I'd like to propose for the source of truth to the BSP to be written in the [Smithy IDL](https://smithy.io/2.0/index.html).

### What is Smithy ?

Smithy is a protocol-agnostic IDL created by AWS. AWS uses Smithy as an input to code-generate all of the publicly available SDKs, as well as server-side bindings for the thousands of service they maintain in their platform. It has two syntaxes:

* the .smithy syntax, exemplified above
* a json syntax, which you can find [examples](https://github.com/aws/aws-sdk-js-v3/tree/main/codegen/sdk-codegen/aws-models) of in the AWS SDK repositories

The smithy metamodel (the building blocks used to construct smithy models) is really simple and intuitive, which makes it ideal as an input to code-generators.

Smithy comes with a number of fundamental building blocks to define data types and interfaces, in particular :

* structures
* unions
* collections (list/maps)
* a bunch of primitive types, including a `document` primitive representing raw json.
* operations (modelled as input/output/errors)
* services (collection of operations)

Additionally, Smithy decouples itself from protocol specific semantics via a powerful annotation mechanism that give meaning to the fundamental building blocks in specific protocols. For instance, it provides a bunch of [http specific](https://smithy.io/2.0/spec/http-bindings.html?highlight=http%20bindings) annotations.

### The Smithy tooling :

Smithy comes with a parser/validator written in Java : [smithy-model](https://search.maven.org/artifact/software.amazon.smithy/smithy-model/1.27.0/jar). It is an extremely well written Java library that makes it really easy to load smithy files and manipulate the models.

Smithy has pretty decent editor support : my team at Disney Streaming maintains a [vscode extension](https://github.com/disneystreaming/vscode-smithy/) for Smithy, and developed the initial [smithy-language-server](https://github.com/awslabs/smithy-language-server) before donating it to AWS. AWS is working on an Intellij plugin but I'm not quite certain what the status is on that.

It also has Github syntax-highlighting.

###  What would be the point ?

Maintaining LSP/BSP servers and clients is hard, there are lots of concerns to take into consideration that wrt concurrency, state, cancellation, timeouts, and graceful error handling ... having the option to generate bindings that integrate tightly with libraries that do this better than what the Java/Scala standard libraries have to offer is immensely helpful. In the current state, it'd be really hard to generate such bindings, and though its existence is great in itself, the `bsp4j` library does not offer the best developer experience compared to the state of the art in the Scala ecosystem(s).

Additionally, using an IDL would drastically lower the bar to create BSP bindings for other languages. AWS is actively working on providing rest/json specific bindings for [rust](https://github.com/awslabs/smithy-rs), [kotlin](https://github.com/awslabs/smithy-kotlin) and [typescript](https://github.com/awslabs/smithy-typescript) and a fair amount of the code could be borrowed from these to code-generate BSP-specific bindings.

As for pure-Scala, the code generated by [Smithy4s](https://github.com/disneystreaming/smithy4s/) is protocol agnostic, and could be used as a basis, but people could also quite easily generate BSP-specific pure-Scala bindings without Smithy4s, just basing themselves on the `smithy-model` library like we do in Smithy4s.

### But what about the current/future state ?

Writing code-generators that use Smithy as an input is a lot easier than writing code-generators that use json-schema or openapi or proto

* The existing xtend templates present in this repo could easily be regenerated from Smithy.
* The typescript "specification" could be generated from Smithy.
* AWS is opening a number of tools (https://github.com/awslabs/smithy-rs)

## Actionable points

If this proposal is even remotely interesting, I could take some time to exemplify what (a subset of) the BSP definition would look like in Smithy to help the discussion, and provide a scala-cli script exemplifying how this definition can be loaded and traversed to retrieve the various shapes.

## Finally

To be absolutely clear: I'm not trying to sell the `smithy4s` library, but from experience, Smithy is the most workable IDL out there for usecases such as the BSP.

**Comment by jastice** [comment]
Thanks for this proposal! We've been considering ways to put the specification on a firmer footing, especially now that we want to expand the protocol implementations to more languages.

So just to understand better, how would you use Smithy to implement the protocol for another language?

**Comment by Baccata** [comment]
> So just to understand better, how would you use Smithy to implement the protocol for another language? 

I'm not quite sure I understand the question, so I'll try to extrapolate. To reiterate, Smithy is an interface definition language, and makes it pretty easy to write code-generators on top of it, it's not meant to "implement" the protocol, directly anyway. It's meant to codify communication in arbitrary protocols. 

I'm gonna try to exemplify what the whole journey would look like 

#### 1. Defining the semantics of the jsonRPC protocol in Smithy. 

So, from the perspective of the engineers defining the BSP, the process would start by defining the semantics of the jsonRPC protocol. The "BSP", in the context of Smithy, is not so much a protocol as it is a "service" that contains operations, and you need to tie each operation to a name that will be used during the runtime interaction (which is a jsonRPC level concern). In order to do that, we codify the semantics of the jsonRPC interaction by defining a bunch of annotations, called "traits". Those are defined as simple datatypes, with the `@trait` annotation to express that they are annotations themselves. 

```smithy 
namespace jsonrpc

@protocolDefinition(traits: [jsonRequest, jsonNotification])
@trait(selector: "service")
structure jsonRPC {}

@trait(selector: "operation")
string jsonRequest
@trait(selector: "operation")
string jsonNotification
```

#### 2. Defining the datatypes that are involved in the BSP. 

To give you an example, you'd model the `Position` datatype as such : 

```smithy
namespace bsp

structure Position {
   @required
   line: Integer
   @required
   character: Integer 
} 

structure Range {
  @required
  start: Position
  @required
  end: Position
} 

string URI 

structure Location {
   @required 
   uri: URI
   @required
   range: Range
}
```

#### 3. Defining the operations and interfaces involved in the BSP

then something like `BuildServer` would look like 

```smithy
namespace bsp

use jsonrpc#jsonRPC
use jsonrpc#jsonRequest
use jsonrpc#jsonNotification

@jsonRPC 
service BuildServer {
   operations: [
      buildInitialize, 
      onBuildInitialize, 
      buildShutdown,
      // ... 
   ]
}

@jsonRequest("build/initialize")
operation BuildInitialize {
   input: InitializeBuildParams
   output: InitializeBuildResult
}

@jsonNotification("build/initialized")
operation OnBuildInitialize {
} 

// ... 
```

#### 4. Implementing the code-generation 

At this point, you have a bunch of `.smithy` files that you can keep in a github repo but also publish to maven in jars. 

Then, you'll obviously need to devise some code-generation pipeline that would take these files and output code conceptually similar to what you have in `bsp4j`. You'd basically identify whether the target language has some existing Json RPC library, and how to use it idiomatically, and then write a code-generator to produce idiomatic bindings (ie server/client stubs) in that language, which would call upon the Json RPC library in question. 

For instance, taking the example of Scala, you could go from the `Position` shape describe above and generate something like

```scala
case class Position(line: Int, character: Int) derives JsonCodec
```

Obviously the service/operations would be more complicated, but you get the idea. 

I presume the most straightforward way to produce the code-generator would be to implement this [interface](https://github.com/awslabs/smithy/blob/96882ea7af0ea5430f11d50fe9840e941e03adf4/smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegen.java#L35), which I admit I've never done : smithy4s (the library I maintain) taps into a lower-level constructs, in particular the model "assembler" which lets you import/parse/validate smithy files and get the list of all data definitions it contains. 

```scala
//> using lib "software.amazon.smithy:smithy-model:1.27.0"
import software.amazon.smithy.model.Model

val modelString = """|namespace bsp
                     |
                     |structure Position {
                     |  line: Integer,
                     |  character: Integer
                     |}
                     |""".stripMargin

@main
def main(): Unit = {
  // parses and validates that the model is coherent
  val model = Model
    .assembler()
    .addUnparsedModel("bsp.smithy", modelString)
    .assemble()
    .unwrap()

  import scala.jdk.CollectionConverters._
  model
    .shapes()
    .iterator()
    .asScala
    .filter(_.getId().getNamespace() == "bsp")
    .foreach(println)

}
```

#### Conclusion 

I hope this helps understand the proposal : defining the BSP in terms of Smithy would be really (much clearer than a Json-schema, I like to think), and would produce a workable artifact that not only you, but also third-parties, could use to code-generate BSP bindings in accordance to their personal preferences in terms of libraries and patterns.

**Comment by ckipp01** [comment]
Hey @Baccata! Thanks a ton for taking the time and writing all this up! I've actually been watching Smithy for a while now ever since I saw you post about Smithy4s, so this is interesting to see!

I can't really speak too in depth since I've _never actually used Smithy before_, but from my understanding I think a change like this could improve quite a few things. I'll also be up front and say that don't really have any love for xtend.

- I know we've spoken I believe on Discord about bindings, like Rust. From poking around I see [smithy-rs](https://github.com/awslabs/smithy-rs) which seems maintained and somewhat popular, but do you have any experience with it? Mainly I'd like to know how good the generation story is outside of Scala.
- From looking at the list of generators in [here](https://smithy.io/2.0/implementations.html#client-code-generators) I don't see a Java one unless I'm totally missing it? I do sort of think that keeping a Java implementation around might be important _unless_ it's really not going to be used much. I know @lefou uses it in Mill.
- Speaking of @lefou, I actually just saw that in our Discord he asked the following:

> bsp4j is currently some (theoretically) lightweight Java API plus some more heavyweight runtime. Is there any chance to consume the pure protocol classes (the assumed lightweight part) without the heavy dependencies like guava ?
> Or asked differently, is there a known set of dependencies, which can be excluded, if I only want to use the API (ch.epfl.scala.bsp4j), but not want to start the server (org.eclipse.lsp4j.jsonrpc)?

This repo is _somewhat_ opinionated on the deps that it brings in, so I'm assuming that if we had some idl like this in theory users could be free to choose to use just the core data structures that are generated from it or the actual services. I'm assuming this type of change could really help in this situation.

A total aside, it's also rad to see that there is LSP support for Smithy and also syntax highlighting on GitHub. Something very noticeably often missing for these sorts of things.

One more question.

> Writing code-generators that use Smithy as an input is a lot easier than writing code-generators that use json-schema or openapi or proto

> defining the BSP in terms of Smithy would be really (much clearer than a Json-schema, I like to think

In regards to a Json-schema, since before this proposal that was the direction I was thinking. Can you maybe outline a bit more some of the concrete advantages of Smithy over this? You hit slightly on a couple, but is the killer feature more that the code generation story is much better, or the actual process of defining BSP is cleaner when using Smithy over  Json-schema? Any other advantages you might be able to think of there I know may help in a decision like this.

Again, thanks for taking the time and writing this all up!

**Comment by jastice** [comment]
> I'm gonna try to exemplify what the whole journey would look like

Thanks, that was exactly what my question was getting at! Thank you for the detailed proposal and explanation so far, and it looks a lot like what we have been wanting. So if some other maintainers agree, I'd love to see more work on this.

To understand better your stake in this, what level of involvement are you thinking of here? Just the bootstrapping of initial structure, or more continued maintenance and implementation?

**Comment by abrams27** [comment]
niceeee, thanks! we've been thinking about doing something similar but using protobuf. In addition, we thought about generating online documentation using this representation (to have only one real source of truth), what do you think?

**Comment by Baccata** [comment]
> From poking around I see [smithy-rs](https://github.com/awslabs/smithy-rs) which seems maintained and somewhat popular, but do you have any experience with it? Mainly I'd like to know how good the generation story is outside of Scala. 

I do not have experience with `smithy-rs` or any other code-generators provided by AWS for that matter. These are protocol specific, and at $work we have our own protocol (which overlaps greatly with one of AWS's but with slight differences). What I can tell you however is that the AWS-rest-json protocol that `smithy-rs` generates bindings for is arguably more complex than jsonRPC, and requires a lot more logic than a jsonRPC protocol, as you have to handle headers, path parameters, query parameters, http bodies, etc ... in comparison, jsonRPC is pretty simple : when invoking a method, the whole input is gonna be translated to json and sent over the wire. My point when linking `smithy-rs` is more that whoever wants to implement RUST bindings could have a point of reference for the code-generation, thus lowering the barrier of entry. 

> From looking at the list of generators in [here](https://smithy.io/2.0/implementations.html#client-code-generators) I don't see a Java one unless I'm totally missing it? I do sort of think that keeping a Java implementation around might be important unless it's really not going to be used much. I know @lefou uses it in Mill. 

I think AWS maintains Java code-generators internally, but in the open-source ecosystem, the kotlin one is probably the closest you'd get. Once again, the code-generators provided by AWS are protocol-specific, and you wouldn't be able to just use them to produce jsonRPC-compatible bindings. 

A pathway that we use internally at $work is to go from Smithy to other IDLs to gain access to the ecosystems surrounding those other IDLs. This approach is facilitated by the fact that AWS provides building blocks to go from Smithy to Openapi, for instance, or from Smithy to Json schema. So if you know of a good code-generator that uses json-schema (or even LSP-flavoured json-schemas) as an input to produce bindings for a language, it'd a worthwhile approach. 

> In regards to a Json-schema, since before this proposal that was the direction I was thinking. Can you maybe outline a bit more some of the concrete advantages of Smithy over this? You hit slightly on a couple, but is the killer feature more that the code generation story is much better, or the actual process of defining BSP is cleaner when using Smithy over Json-schema? Any other advantages you might be able to think of there I know may help in a decision like this.

There is a large list of reasons I could give you that makes Smithy a better IDL than json-schema : 

* readability : you can open a smithy files and understand what the data is gonna look like fairly easily 
* validation tooling : the smithy-model library (which is used to load and validated smithy files) is a lot more robust than anything I've encountered in the json-schema/open-api space. When you load a smithy file using this library, you 100% know the model is correct, which saves you a bunch of null-checks and manual dereferencing 
* Smithy doesn't allow anonymous types : everything has to be named. When writing code-generators for strongly-typed languages, it makes it insanely easier. In comparison, when writing code-generators against openapi and json-schema, you often have to infer names of things based on their locations
* Smithy doesn't allow for nested definitions. Everything is flattened, which makes it a LOT easier for code-generators, once again, as you can do a linear traversal of the list of definitions : `model.shapes().iterator().foreach(renderCode)`. In comparison, both json-schema/openapi allow for nested definitions, which make it a total PITA as you have to traverse a tree instead of a collection. That combined with the fact that nested definitions are often anonymous, it means that you need to infer the name of things based on their positions, which tends to be a lot more complicated than you'd think. 
* `the actual process of defining BSP is cleaner when using Smithy over Json-schema` is very much true, I think 
* Smithy makes it quite easy to load smithy-models from jars : if you publish the smithy files to maven as jars, it means that downstream code-gen pipelines can use tools such as scala-steward to receive automatic updates, it saves a lot of hassle. 
* a pluggable validation mechanism : when you define protocols in smithy (like the jsonRPC I exemplified above), you can write pluggable validators to ensure that the annotations are used properly. For instance, you could come up with custom linting rules that make sure that under a single interface, two operations annotated with `jsonRequest` do not have the same jsonRPC method id. 
* AWS provides a `smithy-diff` library out of the box, which you can use to compare models. It's **really** simple to use, and you could use it to define compatibility rules and run them in CI to ensure that the modifications to the BSP aren't breaking existing tools 

> To understand better your stake in this, what level of involvement are you thinking of here? Just the bootstrapping of initial structure, or more continued maintenance and implementation? 

That is a good question. I'm definitely willing help with the initial bootstrapping, share knowledge, help with writing a code-generator to ensure that the bindings in bsp4j can be re-created, and I can certainly be involved in discussions, but I think it's a bit too premature for me to promise a continued level of maintenance for an undefined duration, in particular in light of the many responsibilities I have. My goal with this proposal is more much to empower the existing maintainers with what I think is the right tool for the job, and to open the door for third parties to improve their own BSP stories. Now if it evolves organically towards a more continuous level of maintenance, so be it, but at this point in time it's not something I can promise. 

## More details that I can share : 

### Anecdotal experience indicating how simple smithy is

On three different occasions, I put smithy in the hands of people who didn't have previous exposure to it. They were able to produce working code-generation prototypes covering most things within hours. One of them wasn't a Scala developer, but a mobile developer and used kotlin for the code-generation. 

### Regarding AWS's 

AWS has about 80 engineers working on smithy-related stuff, internally and in the open source, at any point in time, and saying that they use it extensively would be an understatement. 

AWS engineers are also reasonably responsive on issues in their repositories. I'm certain they wouldn't mind giving pointers to people looking at creating their own code generators. 

Here's a [video](https://www.youtube.com/watch?v=3GpZzu4guTE) from the main maintainer, explaining the origin of Smithy (which is the result of 10+ years of iterations at AWS). 

### Regarding the code-generators and maintenance burden : 

In the context of BSP, code-generators don't have to be published as open-source standalone plugins/general-purpose solutions. For instance : if bsp4j is code-generated, you can publish bsp4j without publishing its code-generator. This lowers the maintenance burden drastically, as you don't have to worry about handling edge-cases such as "what if a user defines an operation that collides with the reserved-keywords of the language" : you can defer that to when the need actually arises. 

In comparison, general-purpose code-generators like smithy4s HAVE to handle edge-cases, because there's always a user who will bump into them.

**Comment by Baccata** [comment]
> In addition, we thought about generating online documentation using this representation (to have only one real source of truth), what do you think? 

Yup, smithy is a very good candidate for this kind of thing. Out of the box, you are provided with a [documentation trait](https://smithy.io/2.0/spec/documentation-traits.html?highlight=documentation), and there is syntactic sugar for it : any triple-slashed comment translates to a documentation annotation when you load the model : 

```smithy
/// hello world 
structure Foo {} 
```

is strictly equivalent to 

```smithy
@documentation("hello world")  
structure Foo {} 
```

Also, because it's much easier to manipulate the model than in proto, it tends to be easier to come up with interesting ways of consuming the model to produce artifacts such as documentation. We definitely use it in such a way at $work.

**Comment by abrams27** [comment]
the only advantage of protobuf - we'd get the generators out of the box

**Comment by Baccata** [comment]
> the only advantage of protobuf - we'd get the generators out of the box

You'd get generators that'd be very biased towards the protobuf serialisation format and the grpc protocol. Doesn't mean that you could easily adjust those to a jsonrpc usecase. 

But since we're talking about protobuf, let me call out the fact that team does maintain a project called [smithy-translate](https://github.com/disneystreaming/smithy-translate) that does **best effort** translations of : 

* json-schema => smithy 
* openapi => smithy 
* smithy => protobuf 

So if you do think some existing code-generators from the proto ecosystem are well suited for the jsonRPC/bsp usecase, that could be an avenue to unlocking that ecosystem. 

**NB** :  I'm highlighting "best-effort", cause both json-schema and openapi are **extremely permissive** IDLs and there are likely some edge-cases that we will never be able to absorb in this smithy-translate tool.

**Comment by jastice** [comment]
Overall this looks very good to me, and your offer to help bootstrap smithy for BSP is very much appreciated. My only concern is whether anybody can invest the time right now to get the current implementations fully ported to a smithy-based version. But if you're okay with a prototype sitting around for a while then I'd love to see what it would look like!

**Comment by Baccata** [comment]
Alrighty, I'll get on with it asap. 

The deliverable of this POC will be a bespoke SBT plugin (inlined in this repo) calling upon a custom code generation pipeline that feeds of smithy models and outputs something as close as possible as the existing bsp4j code, as well as the recreation of (at the very least a subset of) the bsp specification, in smithy syntax, and everything wired into the current build to produce a working artifact.

**Comment by Baccata** [comment]
I'm making reasonably good progress on transposing the spec to smithy semantics. I believe at this point I've got about 40% ported over. If you're curious of the progress, you can go have a look there  : 

https://github.com/Baccata/build-server-protocol/tree/smithy-spec/spec/src/main/resources/META-INF/smithy

**Comment by agluszak** [comment]
Omg, thank you @Baccata! All of that seems almost too good to be true! Thank you very much for your contribution!

**Comment by Baccata** [comment]
At this point, the POC [pull-request](https://github.com/build-server-protocol/build-server-protocol/pull/426) is representative enough that it'd be good for the stakeholders to have a preliminary look. 

Considering how fast I'm going, it wouldn't be unrealistic for the full spec to be ported to smithy by the end of next week. The code-generation is already working and recreating xtend templates that are pretty close to the handcrafted thing. 

It'd be also fairly achievable to turn the current markdown spec into a template that'd receive information from smithy by means of **mdoc**. The typescript code could be recreated from the smithy spec as easily as I've recreated the xtend/java code, and the smithy `@tags` annotation would facilitate the grouping of things together. 

I'm pretty confident that it wouldn't need a big push to eliminate most pain points you guys are experience in the maintenance of this repo in a matter of just a few weeks of on/off work.

**Comment by abrams27** [comment]
btw do we want to introduce breaking changes with this change? if yes, then we have a list of things to change

**Comment by agluszak** [comment]
@Baccata thanks! I'll have a closer look next week, I'm really impressed with the speed of your work!

@abrams27 I don't think so, let's make sure we reach feature parity with the current implementation and migrate to the new one first

**Comment by jastice** [comment]
@Baccata incredible progress, thank you so much! 
So if I understand it right, the PR currently generates bsp4j directly? In that case we could just drop xtend.
For generating the bsp4s bindings, would we want to use smithy4s?

**Comment by Baccata** [comment]
> So if I understand it right, the PR currently generates bsp4j directly? In that case we could just drop xtend.

At the time of writing this, I do generate xtend templates when encountering structure shapes, just because it helps for the concision of the code-generation logic (ie I don't have to generate hashcode/getters/setters/equals). Dropping xtend would certainly be possible, but I'd rather it was done in a subsequent unit of work, if it's the same to you. 

> For generating the bsp4s bindings, would we want to use smithy4s? 

It would certainly be possible to generate pure scala bindings via smithy4s, but you wouldn't be able to recreate anything close to the current bsp4s with smithy4s. The generated code in smithy4s is not customisable and the developer experience would be entirely different. 

So I think you'd be better of re-generating bsp4s in a similar way than I do bsp4j in my PR. Which, in all frankness, would take just a few hours of work, now that the bootstrapping has been taken care of. 

In parallel, I'll probably put together a smithy4s interpreter that would use my [jsonrpclib](https://github.com/neandertech/jsonrpclib) library to provide bindings that'd work with cats-effect and fs2, as this is part of my motivation for doing this work. This would allow third parties who want to use cats-effect/fs2 for their implementations to do it in a pretty straightforward manner.

**Comment by Baccata** [comment]
> btw do we want to introduce breaking changes with this change? if yes, then we have a list of things to change

I think it may be the case, in that breaking changes may have to be introduced in the bsp4j library : the handcrafted POJOs may have some inconsistencies in how they are encoded (constructor of parameters vs getters/setters) that would be ill-advised to try and recreate in a code-gen solution. 

For instance, the `BuildTargetCapabilities` java class has a bunch of constructors that I assume have been handcrafted to retain source/binary compatibility. 

After the first generated bsp4j is out, it would totally be possible to evolve the schemas in ways that would generate binary-compatible code, for instance by only accepting new optional fields or fields with default values (ie prohibiting the addition of required fields to datatypes), but it'd be hard to apply some retroactive logic to re-capture every handcrafted detail.

**Comment by Baccata** [comment]
I'm happy to report that the full markdown specification has been ported to Smithy. 

Some things could be made to improve the aesthetics / readability of the smithy spec, such as using inline definitions for inputs/outputs of operations, using mixins to capture common patterns, or even splitting the `bsp.smithy` file into smaller files (one per operation and one for common/re-used shapes or whatever),  but I think maintainers will have to play with the smithy language themselves and come up with some consensus regarding how they want to organise it. 

From the build perspective, splitting the file into smaller files will change absolutely nothing. The notion of "file" disappears when you load the model, only the notion of "namespace" remains (and several files can populate the same namespace, just like Scala/Java, etc) 

There are a few todos left on the PR, mostly around POC'ing the re-generation of the markdown specification.
